ott/core/implicit_differentiation.py

```diff
@@ -143,21 +143,26 @@ def solve(
     derivative = jax.vmap(jax.grad(precond_fun))
 
     n, m = geom.shape
-    # pylint: disable=g-long-lambda
-    vjp_fg = lambda z: app_transport(
-        f, g, z * derivative(marginal_b(f, g)), axis=1
-    ) / geom.epsilon
-    vjp_gf = lambda z: app_transport(
-        f, g, z * derivative(marginal_a(f, g)), axis=0
-    ) / geom.epsilon
-
-    if not self.symmetric:
-      vjp_fgt = lambda z: app_transport(
-          f, g, z, axis=0
-      ) * derivative(marginal_b(f, g)) / geom.epsilon
-      vjp_gft = lambda z: app_transport(
-          f, g, z, axis=1
-      ) * derivative(marginal_a(f, g)) / geom.epsilon
+
+    if self.symmetric:
+      # pylint: disable=g-long-lambda
+      vjp_fg = lambda z: app_transport(
+          f, g, z * derivative(marginal_b(f, g)), axis=1
+      ) / geom.epsilon
+      vjp_gf = lambda z: app_transport(
+          f, g, z * derivative(marginal_a(f, g)), axis=0
+      ) / geom.epsilon
+
+      solver_fun = jax.scipy.sparse.linalg.cg if self.solver_fun is None else self.solver_fun
+    else:
+      transport = geom.transport_from_potentials(f=f, g=g)
```

Comment:
thanks Ersi! I think you found the problem, the sparse linear solver needs to transpose some of the lambda functions that use `app_transport`. For some reason, this does not work, probably because it cannot transpose the application of transport to a vector. The fix you propose indicates that this transposition can be done efficiently with a matrix, but not with that function.

I do think, however, that instantiating the transport matrix runs against a founding principle of Sinkhorn as defined  in OTT (I know that in 90% of the cases we instantiate the cost/kernel matrices, and therefore in 90% of the cases this would not be a problem, but I am concerned about the remaining 10% :) which include for instance online mode, or grid geometries).

I think a possible and interesting workaround to explore, is to look at custom transposition rules, as introduce by @froystig in this PR https://github.com/google/jax/pull/9134 
it seems that if we are able to define a custom transposition rule for `app_transport` we might finally solve this problem and allow for the use of `gmres` or `bicgstab`. WDYT?

Reply:
Thanks Marco. Yes, I believe that you are right. While the error only appears in the case where ``lse_mode=True``, the issue does not seem to arise from the ``logsumexp`` function itself. I looked into the points discussed last week and I think that in fact we have to go for a custom transposition (most likely
) of ``apply_transport_from_potentials``.